[docs] refresh dev cheat sheet (#24063)
MystenLabs/sui: docs/content/guides/developer/dev-cheat-sheet.mdx

### General

- Read the [Code Quality Checklist](https://move-book.com/guides/code-quality-checklist/) for best practices in Move development.
- Follow the [Move conventions](../../concepts/sui-move-concepts/conventions.mdx) for consistent naming and coding style.

> - Use `vector`-backed collections (`vector`, `VecSet`, `VecMap`, `PriorityQueue`) with a **known** maximum size of ≤ 1000 items.

Reviewer: ```suggestion
- Use `vector`-backed collections (`vector`, `VecSet`, `VecMap`, `PriorityQueue`) with a known maximum size of ≤ 1000 items.
```

  - Use dynamic field-backed collections (`Table`, `Bag`, `ObjectBag`, `ObjectTable`, `LinkedTable`) for any collection that allows third-party addition, larger collections, and collections of unknown size.

>   - Move objects have a maximum size of 250KB—any attempt to create a larger object leads to an aborted transaction. Ensure that your objects do not have an ever-growing `vector`-backed collection.

Reviewer: ```suggestion
  - Move objects have a maximum size of 250KB. Any attempt to create a larger object leads to an aborted transaction. Ensure that your objects do not have an ever-growing `vector`-backed collection.
```

> - If your function `f` needs a payment in (e.g.) SUI from the caller, use `fun f(payment: Coin<SUI>)` not `fun f(payment: &mut Coin<SUI>, amount: u64)`. This is safer for callers—they know exactly how much they are paying, and do not need to trust `f` to extract the right amount.

Reviewer: ```suggestion
- If your function `f` needs a payment in, for example, SUI from the caller, use `fun f(payment: Coin<SUI>)` not `fun f(payment: &mut Coin<SUI>, amount: u64)`. This is safer for callers; they know exactly how much they are paying, and do not need to trust `f` to extract the right amount.
```

### Composability

- Use the [Sui Object Display](https://move-book.com/programmability/display) to customize how your objects show up in wallets, apps, and explorers.
- Avoid “self-transfers”—whenever possible, return the object from the current function, so it can be used in a different command in a [Programmable Transaction Block](./sui-101/working-with-ptbs.mdx).

### Package Upgrades

- Read about [package upgrades](../../concepts/sui-move-concepts/packages/upgrade.mdx) before publishing your package.
- Packages are immutable, so any published package can be called forever. Use [object versioning](../../concepts/sui-move-concepts/packages/upgrade.mdx#versioned-shared-objects) to prevent older versions from being called.

> - If you upgrade a package `P` to `P'`, other packages and clients that depend on `P` will continue using `P`, not auto-update to `P'`. Both dependent packages and client code must be explicitly updated to point at `P'`.

Reviewer: ```suggestion
- If you upgrade a package `P1` to `P2`, other packages and clients that depend on `P1` will continue using `P1`. They do not auto-update to `P2`. Both dependent packages and client code must be explicitly updated to point at `P2`.
```

  - `public` function signatures cannot be deleted or changed, but `public(friend)` functions can. Use `public(friend)` or private visibility liberally unless you are exposing library functions that will live forever.

>   - It is not possible to delete `struct` types, add new fields (though you can add dynamic fields), or add new [abilities](https://move-book.com/reference/abilities) via an upgrade. Introduce new types carefully—they will live forever!

Reviewer: ```suggestion
  - It is not possible to delete `struct` types. Add new fields (though you can add dynamic fields), or add new [abilities](https://move-book.com/reference/abilities) via an upgrade. Introduce new types carefully as they will live forever.
```
